packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -304,9 +291,74 @@ export class OdspDocumentService implements IDocumentService {
         });
     }
 
+    private clearJoinSessionTimer() {
+        if (this.joinSessionRefreshTimer !== undefined) {
+            clearTimeout(this.joinSessionRefreshTimer);
+            this.joinSessionRefreshTimer = undefined;
+        }
+    }
+
+    private async refreshSessionPeriodically(
+        requestWebsocketTokenFromJoinSession: boolean,
+        response: ISocketStorageDiscovery,
+    ) {
+        for (;;) {
+            const refreshSessionDurationSeconds = response.refreshSessionDurationSeconds;
+            if (refreshSessionDurationSeconds !== undefined) {
+                // 30 seconds is buffer time to refresh the session.
+                const retryAfter = (refreshSessionDurationSeconds * 1000) - 30000;
+                // If retryAfter is negative, don't do anything and let the session disconnect.
+                if (retryAfter < 0) {
+                    return;
```

Comment:
I think we should not get here (often), so maybe logging error telemetry if we got here such that we can re-evaluate assumptions if it happens often?